app/routes/_auth+/onboarding.tsx

```diff
@@ -38,24 +37,30 @@ const OnboardingFormSchema = z
 	.object({
 		username: usernameSchema,
 		name: nameSchema,
-		password: passwordSchema,
-		confirmPassword: passwordSchema,
-		agreeToTermsOfServiceAndPrivacyPolicy: checkboxSchema(
-			'You must agree to the terms of service and privacy policy',
-		),
-		agreeToMailingList: checkboxSchema(),
-		remember: checkboxSchema(),
+		agreeToTermsOfServiceAndPrivacyPolicy: z.boolean({
+			required_error:
+				'You must agree to the terms of service and privacy policy',
+		}),
+		agreeToMailingList: z.boolean().optional(),
+		remember: z.boolean().optional(),
 		redirectTo: z.string().optional(),
 	})
-	.superRefine(({ confirmPassword, password }, ctx) => {
-		if (confirmPassword !== password) {
-			ctx.addIssue({
-				path: ['confirmPassword'],
-				code: 'custom',
-				message: 'The passwords did not match',
+	.and(
+		z
+			.object({
+				password: passwordSchema,
+				confirmPassword: passwordSchema,
 			})
-		}
-	})
+			.superRefine(({ confirmPassword, password }, ctx) => {
+				if (confirmPassword !== password) {
+					ctx.addIssue({
+						path: ['confirmPassword'],
+						code: 'custom',
+						message: 'The passwords did not match',
+					})
+				}
+			}),
```

Comment:
Zod does not run `refine` or `superRefine` until all fields are provided. This makes it run earlier as soon as both `password` and `confirmPassoword` is filled.

Ref: https://github.com/colinhacks/zod/issues/479

Reply:
Oh fantastic. Great idea. Thank you!